src/font_renderer/ct_font_renderer.cpp

```diff
@@ -0,0 +1,140 @@
+    GlyphMetrics getGlyphMetrics(GlyphIndex glyphIdx) const override;
+    void renderGlyph(GlyphIndex glyphIdx, Image& image) const override;
+    virtual ~CtFontRenderer() override;
+private:
+    CTFontRef font;
+};
+
+
+CtFontRenderer::CtFontRenderer(const FontRendererArgs& args)
+{
+    CFStringRef fontName = CFStringCreateWithCString(NULL, (const char*)args.data, kCFStringEncodingUTF8);
+    font = CTFontCreateWithName(fontName, args.pxSize, NULL);
+    CFRelease(fontName);
+    if (!font)
+        throw FontRendererError("CoreText can't init font");
+}
+
+
+FontMetrics CtFontRenderer::getFontMetrics() const
+{
+    FontMetrics metrics;
+
+    metrics.ascender = ceil(CTFontGetAscent(font));
+    metrics.descender = ceil(CTFontGetDescent(font));
+    metrics.lineHeight = ceil(CTFontGetLeading(font) + CTFontGetAscent(font) + CTFontGetDescent(font));
+
+    return metrics;
+}
+
+
+GlyphIndex CtFontRenderer::getGlyphIndex(char32_t cp) const
+{
+    CGGlyph glyph;
+    UniChar ch = (UniChar)cp;
+    if (CTFontGetGlyphsForCharacters(font, &ch, &glyph, 1))
+        return (GlyphIndex)glyph;
+    return 0;
+}
+
+#define CORETEXT_SHIFT_X_POSITION 1
+
+GlyphMetrics CtFontRenderer::getGlyphMetrics(GlyphIndex glyphIdx) const
+{
+    GlyphMetrics glyphMetrics;
+    CGGlyph glyph = (CGGlyph)glyphIdx;
+
+    CGSize advance;
+    CTFontGetAdvancesForGlyphs(font, kCTFontOrientationDefault, &glyph, &advance, 1);
+    glyphMetrics.advance = round(advance.width);
+
+    CGRect boundingRect = CTFontGetBoundingRectsForGlyphs(font, kCTFontOrientationDefault, &glyph, NULL, 1);
+#if CORETEXT_SHIFT_X_POSITION
+    glyphMetrics.size.w = ceil(boundingRect.size.width) + ceil(boundingRect.origin.x);
+#else
+    glyphMetrics.size.w = ceil(boundingRect.size.width);
+#endif
+    glyphMetrics.offset.x = ceil(boundingRect.origin.x);
+
+    glyphMetrics.size.h = ceil(boundingRect.size.height);
+    glyphMetrics.offset.y = ceil(boundingRect.size.height) + ceil(boundingRect.origin.y);
+
+    return glyphMetrics;
+}
+
+
+void CtFontRenderer::renderGlyph(GlyphIndex glyphIdx, Image& image) const
+{
+    if (image.getWidth() == 0 || image.getHeight() == 0)
+        return;
+
+    CGColorSpaceRef colorSpace = CGColorSpaceCreateDeviceGray();
+    CGContextRef context = CGBitmapContextCreate(image.getData(), image.getWidth(), image.getHeight(),
+                                                 8, image.getPitch(), colorSpace, kCGImageAlphaOnly);
+
+    CGFontRef cgFont = CTFontCopyGraphicsFont(font, NULL);
+    CGContextSetFont(context, cgFont);
+    CGContextSetFontSize(context, CTFontGetSize(font));
+    CGContextSetFillColorWithColor(context, CGColorGetConstantColor(kCGColorWhite));
+    CGRect boundingRect = CTFontGetBoundingRectsForGlyphs(font, kCTFontOrientationDefault, (const CGGlyph*)&glyphIdx, NULL, 1);
+    CGPoint position = boundingRect.origin;
+#if CORETEXT_SHIFT_X_POSITION
+    position.x = 0;
+#else
+    position.x = -position.x;
+#endif
+    position.y = -position.y;
```

Comment:
You can as well just set coordinates to 0.

Reply:
`x` is explained in the upper conversation, and `y` should be subtracted, otherwise it's rendered outside of the provided area. dpFontBaker expects that the glyph image will be rendered from (0,0) to (width,height). If we ask CoreText to render a glyph *at* (0,0), it will be placed from (0+x,0+y) to (x+width,y+height).

Reply:
Ah, sorry, I mixed it up with -= operator.

Reply:
The renderGlyph() method fills an image of `GlyphMetrics::size` size, which should be `boundingRect.size` in your case; the size should not include drawing offsets.